### ISSUE TITLE
Failed JIT trace mutates model state via LayerNorm captured mean/variance copyback

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#1353]
## Summary

When `CompiledModelCache<float>.GetOrCompileInference(input, () => model.Predict(input))` fails to trace + compile (see #1352), the failure path's `LazyTensorScope.Dispose() → Realize()` cascade RE-RUNS the LayerNorm lazy callback, which calls `eng.LayerNorm(ci, cg, cb, ce, out var freshMean, out var freshVar)` and **copies the result into `capturedMean` / `capturedVar`** — tensors that may be referenced by `LayerNormalizationLayer<T>`'s running-stats / cached forward fields.

Net effect: a single failed `GetOrCompileInference` attempt **silently corrupts the trained model's parameters** with mean/variance values computed against the lazy-traced input (a different shape from training-time inputs after `NormalizeInputBatchDim`).

## Repro

Following the AiDotNet#1352 repro:

```csharp
var arch = new TransformerArchitecture<float>(/* ... */);
var model = new Transformer<float>(arch, new CategoricalCrossEntropyLoss<float>());

// Per-sample training to a meaningful accuracy state...
TrainPerSample(model, corpus, epochs: 2);
model.SetTrainingMode(false);

// Quality snapshot BEFORE any JIT touch.
float[] logitsBefore = ExtractEvalLogits(model, valSlice);
double top1Before = ComputeTop1(logitsBefore);

// One failed JIT trace.
JitCompilationConfig.Default.ApplyToTensorCodec();
try
{
    new CompiledModelCache<float>().GetOrCompileInference(
        new Tensor<float>([1, ctx]),
        () => model.Predict(new Tensor<float>([1, ctx])));
}
catch (Exception) { /* swallow #1352 */ }

// Quality snapshot AFTER the failed trace.
float[] logitsAfter = ExtractEvalLogits(model, valSlice);
double top1After = ComputeTop1(logitsAfter);

Assert.Equal(top1Before, top1After);  // FAILS: deltas of 1-10pp observed
```

## Root cause

`CpuEngine.LayerNorm` GraphMode branch (`CpuEngine.cs` line ~19491) records a variadic lazy node whose execute callback re-calls `LayerNorm(ci, cg, cb, ce, out var freshMean, out var freshVar)` and then:

```csharp
freshMean.AsSpan().CopyTo(capturedMean.AsWritableSpan());
freshVar.AsSpan().CopyTo(capturedVar.AsWritableSpan());
```

`capturedMean` and `capturedVar` are the `mean` and `variance` out-params from the eager LayerNorm call at trace time. Those tensor references may be:
- Held by `LayerNormalizationLayer<T>._mean` / `._variance` for use in the next backward (training) or for forward consistency.
- Used by downstream ops in the SAME forward (no harm because the eager forward already saw the original values).
- Referenced by the lazy node's `savedState` for backward (intended use).

When `Realize()` re-runs the callback (which only happens when the trace FAILED before `MarkCompiled`), the callback overwrites the original mean/var with values computed under a different input contract (NormalizeInputBatchDim auto-promotion, different batch dim, etc.). The next eager `model.Predict` call reads the corrupted mean/var.

## Discrimination probe

Confirmed by inspecting the eval-time top-1 drift in HarmonicEngine's `Phase_PAPER_A_PathB_JitPlanCache_Sanity` (run 1 vs run 2 against the same seed):

| arm | top-1 | PPL | what differs |
|---|---|---|---|
| FP32 baseline | 10.00% | 255.59 | no JIT touch |
| JIT trace-failure + eager fallback (run 1) | 0.00% | 256.00 | trace ran once + failed |
| JIT trace-failure + eager fallback (run 2) | 1.00% | 255.79 | trace ran once + failed (different seed for shuffle) |

Top-1 drift of 3-10pp on a model that scores ~10% top-1 is comparable to "model effectively reset". The PPL barely moves (~0.1-0.5) because the model is near uniform either way, but the argmaxes shuffle, suggesting the LayerNorm normalization stats changed enough to flip the soft-max winners without destroying the energy distribution.

## Expected behavior

`LazyTensorScope.Realize()` should NOT mutate captured tensors when called as part of failure-path cleanup. Two possible fixes:

1. **Make Realize a no-op when the scope wasn't marked compiled AND an exception is in flight**. Track an `ExceptionDispatchInfo` in the scope and skip Realize when one is present.
2. **Defensive copy in the LayerNorm callback**: copy into a callback-local mean/var tensor and only assign into capturedMean/capturedVar if no exception escapes the callback. The current code unconditionally writes before exit.
3. **Document the contract**: if every Engine op's lazy callback is allowed to mutate trace-time captured tensors during Realize, then trace-time captures MUST be defensive copies (which they're currently not for LayerNorm's mean/variance out-params).

## Related issues

- #1352 — the underlying "Destination is too short" trace failure that exposes this side-effect path.
- #1331 — earlier mean/variance staleness bug in the training (forward+backward) path; the inference-only path's analogous robustness is incomplete.

## Why this matters

Consumers attempting to wire the JIT compiled-replay path manually (because `AiModelBuilder.BuildAsync` is mode-collapsed for some configurations — see #1351) have **no safe way to detect whether their model survives a failed trace attempt**. A try/catch around `GetOrCompileInference` is the documented fallback (per `AiModelBuilder.BuildCompiledPredictFunction`), but that catch can't protect against state mutation that happened inside the call.
